apollo-federation/src/connectors/json_selection/methods/public/filter.rs

```diff
@@ -0,0 +1,280 @@
+                }
+                Some(JSON::Bool(false)) | None => {
+                    // Condition is false or errored, exclude the element
+                }
+                Some(_) => {
+                    // Condition returned a non-boolean value, this is an error
+                    has_non_boolean_error = true;
+                    errors.push(ApplyToError::new(
+                        format!(
+                            "->{} condition must return a boolean value",
+                            method_name.as_ref()
+                        ),
+                        input_path.to_vec(),
+                        method_name.range(),
+                    ));
+                }
+            }
+        }
+
+        if has_non_boolean_error {
+            (None, errors)
+        } else {
+            (Some(JSON::Array(output)), errors)
+        }
+    } else {
+        // For non-array inputs, treat as single-element array
+        // Apply filter condition and return either [value] or []
+        let (condition_result, mut condition_errors) =
+            first_arg.apply_to_path(data, vars, input_path);
+
+        match condition_result {
+            Some(JSON::Bool(true)) => (Some(JSON::Array(vec![data.clone()])), condition_errors),
+            Some(JSON::Bool(false)) => (Some(JSON::Array(vec![])), condition_errors),
+            Some(_) => {
+                // Condition returned a non-boolean value, this is an error
+                condition_errors.push(ApplyToError::new(
+                    format!(
+                        "->{} condition must return a boolean value",
+                        method_name.as_ref()
+                    ),
+                    input_path.to_vec(),
+                    method_name.range(),
+                ));
+                (None, condition_errors)
+            }
+            None => {
+                // Condition errored, errors are already in condition_errors
+                (None, condition_errors)
+            }
+        }
+    }
+}
+
+#[allow(dead_code)] // method type-checking disabled until we add name resolution
+fn filter_shape(
+    method_name: &WithRange<String>,
+    method_args: Option<&MethodArgs>,
+    input_shape: Shape,
+    dollar_shape: Shape,
+    named_var_shapes: &IndexMap<&str, Shape>,
+    source_id: &SourceId,
+) -> Shape {
+    let Some(first_arg) = method_args.and_then(|args| args.args.first()) else {
+        return Shape::error(
+            format!("Method ->{} requires one argument", method_name.as_ref()),
+            method_name.shape_location(source_id),
+        );
+    };
+
+    // Compute the shape of the filter condition argument
+    let condition_shape = first_arg.compute_output_shape(
+        input_shape.clone(),
+        dollar_shape,
+        named_var_shapes,
+        source_id,
+    );
+
+    // Validate that the condition evaluates to a boolean
+    if !matches!(condition_shape.case(), ShapeCase::Bool(_)) {
```

Comment:
```suggestion
    // Validate that the condition evaluates to a boolean or
    // something that could become a boolean
    if !(
        matches!(condition_shape.case(), ShapeCase::Bool(_)) ||
        // This allows Unknown and Name shapes, which can produce boolean
        // values at runtime, without any runtime errors.
        condition_shape.accepts(&Shape::unknown([]) {
```
I worry this may be slightly too restrictive, as the condition might produce a placeholder-like shape that _could_ evaluate to a boolean at runtime, like `Unknown` or a `Name`.